### ISSUE TITLE
Add context action to move string interpolation variables to params

### ISSUE THREAD

**Issue by taori** [olsh/resharper-structured-logging#7]
It would be useful if this option existed, because in the past i have gotten pretty used to writing interpolated strings and only recently learned about structured logging.

**Comment by olsh** [comment]
Hi @taori,

Sounds like a great feature. Let me clarify some questions.
Before fix:
```csharp
_log.Information("User {0} with {1} is logged in to {2}", _user, _context.Permissions);
```
After fix:
```csharp
_log.Information("User {User} with {ContextPermissions}", _user, _context.Permissions);
```

Am I correct?

If yes, then we need to figure out what to do with more complicated cases. Here are a couple examples:
```csharp
_log.Information("SomeData {0}", new int[] { 1, 2, 3 });
```
```csharp
_log.Information("Some int {0}", 0);
```
```csharp
_log.Information("Method result {0} {1}", PrintGeneric(user), PrintGeneric(data));
```

Please, let me know what you think.

**Comment by greuelpirat** [comment]
@olsh this would be a great help!

I think the logic should be as simple as possible, to keep it understandable

Unnamed arguments should be just named `Argument`. May be add an additional generic warning for the magic name `Argument`

> ```cs
> _log.Information($"SomeData {new int[] { 1, 2, 3 }}");
> ```
```cs
_log.Information($"SomeData {Argument}", new int[] { 1, 2, 3 });
```

Do not check for constants, add sequence number for duplicates
> ```cs
> _log.Information($"Some int {0}, {1}, {5}");
> ```
```cs
_log.Information($"SomeData {Argument} {Argument1}, {Argument2}", 0, 1, 5);
```

Build names using only characters from expression:
> ```cs
> _log.Information($"Method result {PrintGeneric(user)} {PrintGeneric(data)}");
> ```
```cs
_log.Information($"Method result {PrintGenericUser} {PrintGenericData}", PrintGeneric(user), PrintGeneric(data));
```

What do you think?

**Comment by olsh** [comment]
Hi @greuelpirat 

Thanks for the suggestion! 👍🏻 

> Unnamed arguments should be just named Argument. 

Well, I think this is pretty dangerous if you are using Elasticsearch sink, let's take a look at this code
```csharp
_log.Information($"Another log {"hello"}");
_log.Information($"Some log {42}");
```
The code above will work just fine.

After refactoring that you suggested, the code will look like
```csharp
_log.Information("Another log {Argument}", "hello");
_log.Information("Some log {Argument}", 42);
```
And this is an issue for Elasticsearch
https://github.com/serilog/serilog-sinks-elasticsearch#a-note-about-fields-inside-elasticsearch
> Be aware that there is an explicit and implicit mapping of types inside an Elasticsearch index. A value called X as a string will be indexed as being a string. Sending the same X as an integer in a next log message will not work.

So, I'm not sure that generating constant names is a good solution.

**Comment by vilinski** [comment]
Hi, great suggestion for the string interpolation fix!
The workaround for Elasticsearch would be letting the user to decide how to name the arguments during refactoring. Like it will be done with "introduce variable" refactoring. Maybe the mechanism could be reused here? Resharper is very smart in naming things.
Otherwise, I think it's a problem of Elasticsearch. We use it too, but I have learned about this problem first from your comment :)

**Comment by pfeigl** [comment]
I'd go the following route. Makes it easy for the default use-cases and pretty user-friendly and flexible for the not so easy cases:
* Put structured loggings names based on variable names in `PascalCase` where applicable
* Ask the user for the variable name using something like this in more complex situations (very minimalistic mockup from the rename variable dialog)
![image](https://user-images.githubusercontent.com/237891/164223318-e76707ac-9177-4342-a636-fb6d31f48a49.png)

Optionally, the dialog could be prefilled with something meaningful where possible (i.e. the `PrintGeneric(user)` case)

**Comment by olsh** [comment]
The feature was implemented in #65 and will be available in the next version soon.
Thanks to @matkoch 🔝

**Comment by olsh** [comment]
Closing this out due to inactivity. 
Feel free to re-open the issue.
